contracts/governance/FeeSharingCollector/FeeSharingCollector.sol

```diff
@@ -393,35 +405,83 @@ contract FeeSharingCollector is
      *
      * @dev WARNING! This function skips all the checkpoints before '_fromCheckpoint' irreversibly, use with care
      *
-     * @param _token RBTC dummy to fit into existing data structure or SOV. Former address of the pool token.
-     * @param _fromCheckpoint Skips all the checkpoints before '_fromCheckpoint'
+     * @param _tokens RBTC dummy to fit into existing data structure or SOV. Former address of the pool token.
+     * @param _fromCheckpoints Skips all the checkpoints before '_fromCheckpoint'
      *        should be calculated offchain with getNextPositiveUserCheckpoint function
      * @param _maxCheckpoints Maximum number of checkpoints to be processed.
      * @param _receiver The receiver of tokens or msg.sender
      * */
-    function withdrawStartingFromCheckpoint(
+    function withdrawStartingFromCheckpoints(
+        address[] calldata _tokens,
+        uint256[] calldata _fromCheckpoints,
+        uint32 _maxCheckpoints,
+        address _receiver
+    ) external nonReentrant {
+        validFromCheckpointParams(_fromCheckpoints, _tokens, msg.sender);
+
+        IWrbtcERC20 wrbtcToken = protocol.wrbtcToken();
+        address loanPoolTokenWRBTC = _getAndValidateLoanPoolWRBTC(address(wrbtcToken));
+
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            if (_maxCheckpoints == 0) break;
+            uint256 _fromCheckpoint = _fromCheckpoints[i];
+            address _token = _tokens[i];
+            uint256 endToken;
+
+            uint256 previousProcessedUserCheckpoints = processedCheckpoints[msg.sender][_token];
+            uint256 startingCheckpoint =
+                _fromCheckpoint > previousProcessedUserCheckpoints
+                    ? _fromCheckpoint
+                    : previousProcessedUserCheckpoints;
+
+            if (
```

Comment:
RBTC is sent inside `_withdrawRbtcToken()` function which is reused by another external function, changing this will affect the entire rbtc-based withdrawal related function (which we has combined into non-rbtc withdrawal as well). So it will mostly need huge modifications with the current design functionality. Not sure if this is something we need to cover in this PR?

Reply:
this is an optimization task  
i don't see a reason to break previously achieved optimization - sending only once RBTC tokens extracted from rbtc-based tokens  
solution does require a small refactoring to separate tokens withdrawal from sending to the user  
wdyt @ororopickpocket?

Reply:
updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/507/commits/d88f44fa441ae08c4fc117f5a169596574d95911)